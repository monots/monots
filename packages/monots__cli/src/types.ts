import type { BaseContext } from 'clipanion';

export interface CommandContext extends BaseContext {
  /**
   * The current working directory.
   */
  cwd: string;

  /**
   * The CLI version
   */
  version: string;

  /**
   * The CLI description
   */
  description: string;

  /**
   * The cli name as defined in the package.json.
   */
  name: string;

  /**
   * - `false` when this command is being run directly from the command line
   * - `true` when it is run from another command.
   *
   * @defaultValue false
   */
  internal: boolean;
}

export interface BaseCommandProps {
  verbose?: boolean;
  cwd?: string;
}
