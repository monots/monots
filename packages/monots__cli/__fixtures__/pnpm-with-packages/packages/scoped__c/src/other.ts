export default function () {
  return 'other';
}
