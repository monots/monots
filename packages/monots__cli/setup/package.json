{
  "main": "../dist/setup.cjs.js",
  "module": "../dist/setup.esm.js",
  "types": "../dist/setup.d.ts"
}
