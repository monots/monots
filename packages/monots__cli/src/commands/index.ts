export * from './base-command';
export * from './build-command';
export * from './check-command';
export * from './fix-command';
export * from './init-command';
export * from './prepare-command';
