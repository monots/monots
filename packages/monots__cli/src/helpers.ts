import path from 'node:path';
import url from 'node:url';
import { readPackageUpSync } from 'read-pkg-up';
import updateNotifier from 'update-notifier';

import type { CommandContext } from './types.js';

const SEPARATOR = '__';
const __dirname = path.dirname(url.fileURLToPath(import.meta.url));

/**
 * Notify the user of available updates to the CLI.
 */
export function notifyUpdate(context: CommandContext) {
  const { name, internal, version } = context;

  if (internal) {
    return;
  }

  updateNotifier({ pkg: { name, version } }).notify();
}

export function getPackageJson() {
  const packageJson = readPackageUpSync({ cwd: __dirname })?.packageJson;

  if (!packageJson) {
    throw new Error('Invalid installation of `monots`');
  }

  return packageJson;
}

/**
 * Get the absolute path within this package.
 */
export function getPackagePath(...paths: string[]) {
  return path.join(__dirname, '..', ...paths);
}

/**
 * Convert a mangled name to its unmangled version.
 *
 * `babel__types` => `@babel/types`.
 */
export function unmangleScopedPackage(mangledName: string): string {
  return mangledName.includes(SEPARATOR) ? `@${mangledName.replace(SEPARATOR, '/')}` : mangledName;
}

/**
 * Mangle a scoped package name. Which removes the `@` symbol and adds a `__`
 * separator.
 *
 * `@babel/types` => `babel__types`
 */
export function mangleScopedPackageName(packageName: string): string {
  if (packageName.indexOf('@') === 0 && packageName.includes('/')) {
    // we have a scoped module, e.g. @bla/foo
    // which should be converted to   bla__foo
    packageName = packageName.slice(1).replace('/', '__');
  }

  return packageName;
}
