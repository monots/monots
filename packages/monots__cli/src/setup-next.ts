import { loadConfig, NAME } from '@monots/core';
import type { CommandContext } from '@monots/types';
import { getPackageJsonSync } from '@monots/utils';
import { Builtins, Cli } from 'clipanion';
import parser from 'yargs-parser';

const { version, description = '', name } = getPackageJsonSync();

/**
 * Loads the `monots` configuration and uses it to register all commands.
 */
export async function createCli() {
  const argv = parser(process.argv.slice(2));
  const logLevel = argv.logLevel || argv.l;
  const result = await loadConfig();

  if (!result) {
    throw new Error('No configuration found for `monots`');
  }

  const cli = new Cli<CommandContext>({
    binaryLabel: description,
    binaryName: NAME,
    binaryVersion: version,
  });

  cli.register(Builtins.HelpCommand);
  cli.register(Builtins.VersionCommand);

  // Gather commands
}

export const context: CommandContext = {
  internal: false,
  version,
  description,
  name,
  cwd: process.cwd(),
  stdin: process.stdin,
  stdout: process.stdout,
  stderr: process.stderr,
};
