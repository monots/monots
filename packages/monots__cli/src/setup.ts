import { NAME } from '@monots/core';
import type { CommandContext } from '@monots/types';
import { getPackageJsonSync } from '@monots/utils';
import { Builtins, Cli } from 'clipanion';

import {
  BuildCommand,
  CheckCommand,
  CreateCommand,
  FixCommand,
  InitCommand,
  PrepareCommand,
} from './commands/index.js';

const { version, description = '', name } = getPackageJsonSync();

export const cli = new Cli<CommandContext>({
  binaryLabel: description,
  binaryName: NAME,
  binaryVersion: version,
});

cli.register(Builtins.HelpCommand);
cli.register(Builtins.VersionCommand);
cli.register(InitCommand);
cli.register(PrepareCommand);
cli.register(FixCommand);
cli.register(BuildCommand);
cli.register(CheckCommand);
cli.register(CreateCommand);

/**
 * Loads the `monots` configuration and uses it to register all commands.
 */
export function registerCommands() {}

export const context: CommandContext = {
  internal: false,
  version,
  description,
  name,
  colorDepth: 256,
  env: process.env,
  cwd: process.cwd(),
  stdin: process.stdin,
  stdout: process.stdout,
  stderr: process.stderr,
};
