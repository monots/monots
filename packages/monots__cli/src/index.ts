#!/usr/bin/env node

import chalk from 'chalk';

import { cli, context } from './setup.js';

cli.runExit(process.argv.slice(2), context).catch((error) => {
  console.log(chalk.red('Unexpected error. Please report it as a bug:'));
  console.log(error);
  console.log();
});
