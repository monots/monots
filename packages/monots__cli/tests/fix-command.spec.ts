import { createSetupFixtures } from '@monots/test';
import { afterAll, test } from 'vitest';

import { cli, context } from '../src/setup';

const setupFixtures = createSetupFixtures({ context, fileUrl: import.meta.url });

afterAll(async () => {
  await setupFixtures.cleanup();
});

test.concurrent('`monots fix` should update the package.json files', async ({ expect }) => {
  const { context, loadJsonFile } = await setupFixtures('pnpm-with-packages');
  const result = await cli.run(['fix'], context);
  const jsonA = await loadJsonFile('packages/scoped__a/package.json');
  const jsonB = await loadJsonFile('packages/scoped__b/package.json');
  const jsonC = await loadJsonFile('packages/scoped__c/package.json');
  const jsonTs = await loadJsonFile('packages/scoped__ts/package.json');

  expect(result, 'The result is successful').toBe(0);
  expect(jsonA).toMatchSnapshot();
  expect(jsonB).toMatchSnapshot();
  expect(jsonC).toMatchSnapshot();
  expect(jsonTs).toMatchSnapshot();
});

test.concurrent('`monots fix` should update the tsconfig files', async ({ expect }) => {
  const { context, loadJsonFile } = await setupFixtures('pnpm-with-packages');
  const result = await cli.run(['fix'], context);
  const jsonA = await loadJsonFile('packages/scoped__a/src/tsconfig.json');
  const jsonB = await loadJsonFile('packages/scoped__b/src/tsconfig.json');
  const jsonC = await loadJsonFile('packages/scoped__c/src/tsconfig.json');
  const jsonTs = await loadJsonFile('packages/scoped__ts/tsconfig.json');

  expect(result, 'The result is successful').toBe(0);
  expect(jsonA).toMatchSnapshot();
  expect(jsonB).toMatchSnapshot();
  expect(jsonC).toMatchSnapshot();
  expect(jsonTs).toMatchSnapshot();
});

test.concurrent(
  '`monots fix` should update the relative baseTsconfig files',
  async ({ expect }) => {
    const { context, loadJsonFile } = await setupFixtures('pnpm-with-packages-relative');
    const result = await cli.run(['fix'], context);
    const jsonA = await loadJsonFile('packages/scoped__a/src/tsconfig.json');
    const jsonB = await loadJsonFile('packages/scoped__b/src/tsconfig.json');

    expect(result, 'The result is successful').toBe(0);
    expect(jsonA).toMatchSnapshot();
    expect(jsonB).toMatchSnapshot();
  },
);

test.concurrent('`monots fix` should support ignoring exports', async ({ expect }) => {
  const { context, loadJsonFile } = await setupFixtures('pnpm-ignore-exports');
  await cli.run(['fix'], context);
  const jsonA = await loadJsonFile('packages/scoped__a/package.json');

  expect(jsonA).toMatchSnapshot();
});

test.concurrent('`monots fix` should add exports to entrypoints', async ({ expect }) => {
  const { context, loadJsonFile } = await setupFixtures('pnpm-with-add-exports-to-entrypoints');
  await cli.run(['fix'], context);
  const mainJson = await loadJsonFile('packages/scoped__add/package.json');
  const scopedJson = await loadJsonFile('packages/scoped__add/other/package.json');

  expect(mainJson).toMatchSnapshot();
  expect(scopedJson).toMatchSnapshot();
});
