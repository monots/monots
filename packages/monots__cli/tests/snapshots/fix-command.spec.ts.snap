// Vitest Snapshot v1

exports[`\`monots fix\` should add exports to entrypoints 1`] = `
{
  "browser": "./dist/index.browser.esm.js",
  "exports": {
    ".": {
      "browser": "./dist/index.browser.esm.js",
      "import": "./dist/index.esm.js",
      "require": "./dist/index.cjs.js",
      "types": "./dist/index.d.ts",
    },
    "./extra": "./extra.js",
    "./index.js": {
      "browser": "./dist/index.browser.esm.js",
      "import": "./dist/index.esm.js",
      "require": "./dist/index.cjs.js",
      "types": "./dist/index.d.ts",
    },
    "./other": {
      "browser": "./dist/other.browser.esm.js",
      "import": "./dist/other.esm.js",
      "require": "./dist/other.cjs.js",
      "types": "./dist/other.d.ts",
    },
    "./other.js": {
      "browser": "./dist/other.browser.esm.js",
      "import": "./dist/other.esm.js",
      "require": "./dist/other.cjs.js",
      "types": "./dist/other.d.ts",
    },
    "./package.json": "./package.json",
    "./types/*": "./dist/*.d.ts",
  },
  "files": [
    "dist",
    "other",
  ],
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "monots": {
    "addExportsToEntrypoints": true,
    "entrypoints": [
      "index.ts",
      "other.ts",
    ],
    "externalModules": [],
    "extraExports": {
      "./extra": "./extra.js",
    },
    "mode": "library",
    "tsconfigs": {},
  },
  "name": "@scoped/add",
  "type": "module",
  "types": "./dist/index.d.ts",
  "version": "1.0.0",
}
`;

exports[`\`monots fix\` should add exports to entrypoints 2`] = `
{
  "browser": "../dist/other.browser.esm.js",
  "exports": {
    ".": {
      "browser": "./../dist/other.browser.esm.js",
      "import": "./../dist/other.esm.js",
      "require": "./../dist/other.cjs.js",
      "types": "./../dist/other.d.ts",
    },
    "./index.js": {
      "browser": "./../dist/other.browser.esm.js",
      "import": "./../dist/other.esm.js",
      "require": "./../dist/other.cjs.js",
      "types": "./../dist/other.d.ts",
    },
  },
  "main": "../dist/other.cjs.js",
  "module": "../dist/other.esm.js",
  "types": "../dist/other.d.ts",
}
`;

exports[`\`monots fix\` should support ignoring exports 1`] = `
{
  "browser": "./dist/index.browser.esm.js",
  "dependencies": {
    "@scoped/b": "^2.0.0",
  },
  "files": [
    "dist",
  ],
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "monots": {
    "entrypoints": [
      "index.{ts,tsx}",
    ],
    "externalModules": [],
    "ignoreExports": true,
    "mode": "library",
    "tsconfigs": {},
  },
  "name": "@scoped/a",
  "type": "module",
  "types": "./dist/index.d.ts",
  "version": "1.0.0",
}
`;

exports[`\`monots fix\` should update the package.json files 1`] = `
{
  "browser": "./dist/index.browser.esm.js",
  "dependencies": {
    "@scoped/b": "^2.0.0",
  },
  "exports": {
    ".": {
      "browser": "./dist/index.browser.esm.js",
      "import": "./dist/index.esm.js",
      "require": "./dist/index.cjs.js",
      "types": "./dist/index.d.ts",
    },
    "./index.js": {
      "browser": "./dist/index.browser.esm.js",
      "import": "./dist/index.esm.js",
      "require": "./dist/index.cjs.js",
      "types": "./dist/index.d.ts",
    },
    "./package.json": "./package.json",
    "./types/*": "./dist/*.d.ts",
  },
  "files": [
    "dist",
  ],
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "name": "@scoped/a",
  "type": "module",
  "types": "./dist/index.d.ts",
  "version": "1.0.0",
}
`;

exports[`\`monots fix\` should update the package.json files 2`] = `
{
  "browser": "./dist/index.browser.esm.js",
  "dependencies": {
    "@scoped/c": "^1.0.0",
  },
  "exports": {
    ".": {
      "browser": "./dist/index.browser.esm.js",
      "import": "./dist/index.esm.js",
      "require": "./dist/index.cjs.js",
    },
    "./index.js": {
      "browser": "./dist/index.browser.esm.js",
      "import": "./dist/index.esm.js",
      "require": "./dist/index.cjs.js",
    },
    "./package.json": "./package.json",
  },
  "files": [
    "dist",
  ],
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "name": "@scoped/b",
  "type": "module",
  "version": "2.0.0",
}
`;

exports[`\`monots fix\` should update the package.json files 3`] = `
{
  "browser": "./dist/index.browser.esm.js",
  "exports": {
    ".": {
      "browser": "./dist/index.browser.esm.js",
      "import": "./dist/index.esm.js",
      "require": "./dist/index.cjs.js",
      "types": "./dist/index.d.ts",
    },
    "./and-another": {
      "browser": "./dist/and-another.browser.esm.js",
      "import": "./dist/and-another.esm.js",
      "require": "./dist/and-another.cjs.js",
      "types": "./dist/and-another.d.ts",
    },
    "./and-another.js": {
      "browser": "./dist/and-another.browser.esm.js",
      "import": "./dist/and-another.esm.js",
      "require": "./dist/and-another.cjs.js",
      "types": "./dist/and-another.d.ts",
    },
    "./and-another/nested": {
      "browser": "./dist/and-another/nested.browser.esm.js",
      "import": "./dist/and-another/nested.esm.js",
      "require": "./dist/and-another/nested.cjs.js",
      "types": "./dist/and-another/nested.d.ts",
    },
    "./and-another/nested.js": {
      "browser": "./dist/and-another/nested.browser.esm.js",
      "import": "./dist/and-another/nested.esm.js",
      "require": "./dist/and-another/nested.cjs.js",
      "types": "./dist/and-another/nested.d.ts",
    },
    "./and-another/nested/cool": {
      "browser": "./dist/and-another/nested/cool.browser.esm.js",
      "import": "./dist/and-another/nested/cool.esm.js",
      "require": "./dist/and-another/nested/cool.cjs.js",
      "types": "./dist/and-another/nested/cool.d.ts",
    },
    "./and-another/nested/cool.js": {
      "browser": "./dist/and-another/nested/cool.browser.esm.js",
      "import": "./dist/and-another/nested/cool.esm.js",
      "require": "./dist/and-another/nested/cool.cjs.js",
      "types": "./dist/and-another/nested/cool.d.ts",
    },
    "./extra": "./extra.js",
    "./index.js": {
      "browser": "./dist/index.browser.esm.js",
      "import": "./dist/index.esm.js",
      "require": "./dist/index.cjs.js",
      "types": "./dist/index.d.ts",
    },
    "./other": {
      "browser": "./dist/other.browser.esm.js",
      "import": "./dist/other.esm.js",
      "require": "./dist/other.cjs.js",
      "types": "./dist/other.d.ts",
    },
    "./other.js": {
      "browser": "./dist/other.browser.esm.js",
      "import": "./dist/other.esm.js",
      "require": "./dist/other.cjs.js",
      "types": "./dist/other.d.ts",
    },
    "./package.json": "./package.json",
    "./types/*": "./dist/*.d.ts",
  },
  "files": [
    "and-another",
    "cool",
    "dist",
    "nested",
    "other",
  ],
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "monots": {
    "entrypoints": [
      "index.ts",
      "other.ts",
      "and-another/index.ts",
      "and-another/nested/cool.ts",
      "and-another/nested/index.ts",
    ],
    "externalModules": [],
    "extraExports": {
      "./extra": "./extra.js",
    },
    "mode": "library",
    "tsconfigs": {},
  },
  "name": "@scoped/c",
  "type": "module",
  "types": "./dist/index.d.ts",
  "version": "1.0.0",
}
`;

exports[`\`monots fix\` should update the package.json files 4`] = `
{
  "browser": "",
  "dependencies": {
    "@scoped/b": "^2.0.0",
  },
  "exports": {
    "./package.json": "./package.json",
    "./types/*": "./dist/*.d.ts",
  },
  "files": [
    "dist",
  ],
  "main": "",
  "module": "",
  "monots": {
    "entrypoints": [],
    "externalModules": [],
    "mode": "library",
    "sourceFolderName": "",
    "tsconfigs": {},
  },
  "name": "@scoped/ts",
  "type": "module",
  "types": "",
  "version": "1.0.0",
}
`;

exports[`\`monots fix\` should update the relative baseTsconfig files 1`] = `
{
  "__AUTO_GENERATED__": [
    "To update the configuration edit the following field in the 'package.json' for this package.",
    "\`monots.tsconfigs.src\`",
    "",
    "Then run the command to automatically update everything: \`monots fix\`",
  ],
  "compilerOptions": {
    "composite": true,
    "declaration": true,
    "emitDeclarationOnly": true,
    "noEmit": false,
    "outDir": "../dist",
    "types": [],
  },
  "extends": "../../../relative/tsconfig.json",
  "references": [
    {
      "path": "../../scoped__b/src",
    },
  ],
}
`;

exports[`\`monots fix\` should update the relative baseTsconfig files 2`] = `
{
  "__AUTO_GENERATED__": [
    "To update the configuration edit the following field in the 'package.json' for this package.",
    "\`monots.tsconfigs.src\`",
    "",
    "Then run the command to automatically update everything: \`monots fix\`",
  ],
  "compilerOptions": {
    "noEmit": true,
    "types": [],
  },
  "extends": "../../../relative/tsconfig.json",
  "references": [],
}
`;

exports[`\`monots fix\` should update the tsconfig files 1`] = `
{
  "__AUTO_GENERATED__": [
    "To update the configuration edit the following field in the 'package.json' for this package.",
    "\`monots.tsconfigs.src\`",
    "",
    "Then run the command to automatically update everything: \`monots fix\`",
  ],
  "compilerOptions": {
    "composite": true,
    "declaration": true,
    "emitDeclarationOnly": true,
    "noEmit": false,
    "outDir": "../dist",
    "types": [],
  },
  "extends": "@monots/tsconfig/tsconfig.json",
  "references": [
    {
      "path": "../../scoped__b/src",
    },
  ],
}
`;

exports[`\`monots fix\` should update the tsconfig files 2`] = `
{
  "__AUTO_GENERATED__": [
    "To update the configuration edit the following field in the 'package.json' for this package.",
    "\`monots.tsconfigs.src\`",
    "",
    "Then run the command to automatically update everything: \`monots fix\`",
  ],
  "compilerOptions": {
    "noEmit": true,
    "types": [],
  },
  "extends": "@monots/tsconfig/tsconfig.json",
  "references": [
    {
      "path": "../../scoped__c/src",
    },
  ],
}
`;

exports[`\`monots fix\` should update the tsconfig files 3`] = `
{
  "__AUTO_GENERATED__": [
    "To update the configuration edit the following field in the 'package.json' for this package.",
    "\`monots.tsconfigs.src\`",
    "",
    "Then run the command to automatically update everything: \`monots fix\`",
  ],
  "compilerOptions": {
    "composite": true,
    "declaration": true,
    "emitDeclarationOnly": true,
    "noEmit": false,
    "outDir": "../dist",
    "types": [],
  },
  "extends": "@monots/tsconfig/tsconfig.json",
  "references": [],
}
`;

exports[`\`monots fix\` should update the tsconfig files 4`] = `
{
  "__AUTO_GENERATED__": [
    "To update the configuration edit the following field in the 'package.json' for this package.",
    "\`monots.tsconfigs.\`",
    "",
    "Then run the command to automatically update everything: \`monots fix\`",
  ],
  "compilerOptions": {
    "composite": true,
    "declaration": true,
    "emitDeclarationOnly": true,
    "noEmit": false,
    "outDir": "dist",
    "types": [],
  },
  "extends": "@monots/tsconfig/tsconfig.json",
  "references": [
    {
      "path": "../scoped__b/src",
    },
  ],
}
`;
